pkg/ruler/ruler.go

```diff
@@ -722,6 +752,70 @@ func filterRuleGroupByEnabled(group *rulespb.RuleGroupDesc, recordingEnabled, al
 	return filtered, removedRules
 }
 
+// filterRuleGroupsByNotMissing filters out from the input configs all the rules groups which are in the missing list.
+//
+// This function doesn't modify the input configs in place (even if it could) in order to reduce the likelihood of introducing
+// future bugs, in case the rule groups will be cached in memory.
+func filterRuleGroupsByNotMissing(configs map[string]rulespb.RuleGroupList, missing rulespb.RuleGroupList, logger log.Logger) (filtered map[string]rulespb.RuleGroupList) {
+	// Nothing to do if there are no missing rule groups.
+	if len(missing) == 0 {
+		return configs
+	}
+
+	// Build a map to easily lookup missing rule groups.
+	getRuleGroupLookupKey := func(group *rulespb.RuleGroupDesc) string {
+		var sep = string([]byte{0})
+		return group.GetUser() + sep + group.GetNamespace() + sep + group.GetName()
```

Comment:
Some of our rulers have tens of thousands rule groups. On one hand this PR optimizes allocation in cache keys formatting, but this will be quite inefficient too.

(I don't think this needs fixing, until profiler shows otherwise. Just found it interesting.)

Reply:
My rationale (which I admit wasn't clear just reading the PR):
- Missing rule groups are expected to be very rare. Most of the times this filtering function is a no-op.
- The main reason why I wanted to optimize the caching bucket cache key generation was for the blocks storage cache (store-gateway) not actually the ruler 😉